pkg/metrics/pushclient/pushclient_test.go

```diff
@@ -0,0 +1,92 @@
+package pushclient
+
+import (
+	"io"
+	"net/http"
+	"strings"
+	"testing"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stretchr/testify/assert"
+)
+
+// testClient is used as a mock client to test the functionality.
+type testClient struct {
+	// embed a stdlib http.Client
+	t              *testing.T
+	expectedString string
+	*http.Client
+}
+
+// Do is a test function that is used to test the metrics push client.
+func (a testClient) Do(req *http.Request) (resp *http.Response, err error) {
+	buf := new(strings.Builder)
+	_, err = io.Copy(buf, req.Body)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse prometheus collector request")
+	}
+	assert.EqualValues(a.t, a.expectedString, buf.String())
+	return &http.Response{
+		Body:       req.Body,
+		Status:     "OK",
+		StatusCode: 200,
+	}, nil
+}
+
+func TestPushMetricsToGateway(t *testing.T) {
+	promCollector := prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name:        "test",
+			Help:        "test metrics",
+			ConstLabels: map[string]string{"test1": "test1"},
+		},
+	)
+
+	promCollector.Add(10)
+
+	expectedOutput := "# HELP test test metrics\n# TYPE test counter\ntest{test1=\"test1\"} 10\n"
+	err := pushToAggregationGateway(gatewayURL, "test", promCollector, &testClient{t, expectedOutput, &http.Client{}})
+	if err != nil {
+		assert.Failf(t, err.Error(), "error pushing metrics")
+	}
+}
+
+func TestPushMetricsToAllGateway(t *testing.T) {
+	counter := prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name:        "test_counter",
+			Help:        "test metrics",
+			ConstLabels: map[string]string{"test1": "test1"},
+		},
+	)
+	counter.Add(10)
+
+	histogram := prometheus.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:        "test_histogram",
+			Help:        "test metrics",
+			ConstLabels: map[string]string{"test1": "test1"},
+		},
+	)
+	histogram.Observe(10)
+
+	promCollector := []prometheus.Collector{counter, histogram}
+
+	expectedOutput := "# HELP test_counter test metrics\n# " +
+		"TYPE test_counter counter\ntest_counter{test1=\"test1\"} 10\n" +
+		"# HELP test_histogram test metrics\n" +
+		"# TYPE test_histogram histogram\ntest_histogram_bucket{test1=\"test1\",le=\"0.005\"} 0\n" +
+		"test_histogram_bucket{test1=\"test1\",le=\"0.01\"} 0\ntest_histogram_bucket{test1=\"test1\",le=\"0.025\"} 0\n" +
+		"test_histogram_bucket{test1=\"test1\",le=\"0.05\"} 0\ntest_histogram_bucket{test1=\"test1\",le=\"0.1\"} 0\n" +
+		"test_histogram_bucket{test1=\"test1\",le=\"0.25\"} 0\ntest_histogram_bucket{test1=\"test1\",le=\"0.5\"} 0\n" +
+		"test_histogram_bucket{test1=\"test1\",le=\"1\"} 0\ntest_histogram_bucket{test1=\"test1\",le=\"2.5\"} 0\n" +
+		"test_histogram_bucket{test1=\"test1\",le=\"5\"} 0\ntest_histogram_bucket{test1=\"test1\",le=\"10\"} 1\n" +
+		"test_histogram_bucket{test1=\"test1\",le=\"+Inf\"} 1\ntest_histogram_sum{test1=\"test1\"} 10\n" +
+		"test_histogram_count{test1=\"test1\"} 1\n"
```

Comment:
use https://github.com/openshift/installer/pull/3762/files#r441307417 to make this more readable.